README.md
# MultiThreadSocketComsBankSimulation

this here is a project made for the psr subject by ben mansour mahdi
Do note that all of this code is not industry oriented but is instsead made for the sake of experimentation
mainly with multithread programming aka parallel processing and sockets tcp communication over lan 

# Bank Simulation

in the context of multithreading  subject in the second year of engineering in enicarthage 
i (inube) have been assigned this project which is basically a simulation of a local bank agency responsible of keeping 
accounts data including account ribs their solde and status (negative or positive)
more details in the pdf in this repository 



# Notes

## notes and optimization for the future :
+in the future i will add a little bit of coding to make use of the arp algorithms for automatic detection of the open sockets serving this application
from the client side 
+also a better file management method is in order 
+eliminate the os cls methods as they are windows specific 
update the guy / ibuild for linux users and fix the mobile apk app
+update the apk to google drive and add it to this repo



history.txt
1234,Withdrawal,100,Success,Negativ
14522,Deposit,2000,Success,Positiv
156,Withdrawal,2201,Success,Negativ
14522,Deposit,2200,Success,Positiv
156,Withdrawal,100,Success,Negativ
1234,Withdrawal,100,Failure,Negativ
1234,Withdrawal,100,Failure,Negativ
1234,Withdrawal,100,Failure,Negativ
1234,Withdrawal,100,Success,Negativ
1234,Withdrawal,100,Failure,Negativ
100,Deposit,150,Success,Positiv
50,Deposit,150,Success,Positiv
100,Deposit,150,Success,Positiv
1452,Deposit,150,Success,Positiv
1452,Deposit,150,Success,Positiv
1452,Deposit,150,Success,Positiv
1452,Deposit,150,Success,Positiv
1452,Deposit,150,Success,Positiv
1452,Deposit,150,Success,Positiv
1452,Deposit,150,Success,Positiv
1452,Deposit,150,Success,Positiv
1452,Deposit,150,Success,Positiv
1452,Deposit,150,Success,Positiv
1452,Deposit,150,Success,Positiv
1452,Deposit,150,Success,Positiv
1452,Deposit,150,Success,Positiv
1452,Deposit,150,Success,Positiv
1452,Deposit,150,Success,Positiv
1452,Deposit,150,Success,Positiv
1234,Deposit,100,Success,Negativ
1234,Deposit,100,Success,Negativ
1234,Deposit,300,Success,Positiv
1234,Deposit,200,Success,Positiv
1234,Deposit,1000,Success,Positiv
1234,Deposit,500,Success,Positiv
1234,Deposit,500,Success,Positiv
1234,Deposit,500,Success,Positiv
1234,Deposit,12,Success,Positiv
1234,Deposit,500,Success,Positiv
156,Deposit,100,Success,Positiv
156,Withdrawal,500,Success,Positiv
156,Deposit,1000,Success,Positiv
156,Withdrawal,500,Success,Positiv
156,Withdrawal,100,Success,Positiv
156,Deposit,100,Success,Positiv
156,Deposit,500,Success,Positiv
156,Deposit,500,Success,Positiv
156,Withdrawal,0,Success,Positiv
156,Withdrawal,500,Success,Positiv
156,Deposit,500,Success,Positiv
156,Withdrawal,3000,Success,Negativ
156,Deposit,2500,Success,Positiv
156,Withdrawal,3000,Success,Negativ
156,Deposit,490,Success,Negativ
1234,Deposit,1000,Success,Positiv
1234,Withdrawal,4300,Success,Negativ

GUi/main.py
import socket
from kivy.app import App
from kivy.lang import Builder
from kivy.uix.screenmanager import ScreenManager, Screen
from kivy.uix.textinput import TextInput
from kivy.uix.label import Label

# Create both screens. Please note the root.manager.current: this is how
# you can control the ScreenManager from kv. Each screen has by default a
# property manager that gives you the instance of the ScreenManager used.
Builder.load_string(
"""
<inputHnadler>
    id:IM
    name:inputmanager


<ScreenManagement>:

    MenuScreen:
        id: name
        name: 'menu'
        
    SettingsScreen:
        id: settings
        name: 'settings'
    TransactionsScreen:
        id:transactions
        name:'transactions'
    MonitoringScreen:
        id:monitoring
        name:'monitoring'
    InfoScreen
        id:info
        name:'info'



<MonitoringScreen>:
    canvas.before:
   
        Rectangle:
    
            pos: self.pos
            size: self.size
            source: 'cartoon.jpg' 
    FloatLayout:
        
        post:0,0
        size: root.width,root.height
        TextInput:
            id: rib
            multiline:False
            pos_hint: {'center_x': 0.7, 'center_y': 0.705}
            size_hint: 0.2, 0.05
            hint_text:'Rib'
        Button:
            id:balancebutton
            size_hint:.2,.1
            text:'Check Balance'
            pos_hint:{'x':0.1,'y':0.6}
            on_release:  
                root.manager.ids.info.showbalance(rib.text)
                root.manager.current='info'
                
        Button:
            id:transactionsbutton
            size_hint:.2,.1
            text:'View Transactions'
            pos_hint:{'x':0.1,'y':0.4}
            on_release:  
                root.manager.ids.info.showTransactions(rib.text)
               
                
        Button:
            id:debtsbutton
            pos_hint:{'x':0.1,'y':0.2}
            size_hint:.2,.1
            text:'View Bank Debts'
            on_release:   
                root.manager.ids.info.showfacture(rib.text)


        Button:
            pos_hint:{'x':0.8,'y':0.9}
            size_hint:.1,.05
            text:'Home'
            on_release:  
                root.manager.current = 'menu'


<InfoScreen>:
    canvas.before:
        Rectangle:
    
            pos: self.pos
            size: self.size
            source: 'télécharger.png' 
    FloatLayout:
  
        post:0,0
        size: root.width,root.height
        TextInput:
            id: transactions
            pos_hint: {'center_x': 0.5, 'center_y': 0.55}
            size_hint: 0.5, 0.7
            hint_text:'transactions'
        Button:
            pos_hint:{'x':0.8,'y':0.9}
            size_hint:.1,.05
            text:'Home'
            on_release:  
                root.manager.current = 'menu'




<TransactionsScreen>:
    canvas.before:
   
        Rectangle:
    
            pos: self.pos
            size: self.size
            source: 'mymoney.png' 
    FloatLayout:
  
        post:0,0
        size: root.width,root.height
        Button:
            size_hint:.2,.1
            text:'Withdraw'
            pos_hint:{'x':0.08,'y':0.8}
            on_release:  
                root.manager.ids.transactions.withdraw(trrib.text,tramount.text)
        Button:
            size_hint:.2,.1
            text:'Deposit'
            pos_hint:{'x':0.3,'y':0.8}
            on_release: 
                root.manager.ids.transactions.deposit(trrib.text,tramount.text) 
        Button:
            pos_hint:{'x':0.8,'y':0.9}
            size_hint:.1,.05
            text:'Home'
            on_release:  
                root.manager.current = 'menu'
                #TextInput(text='Hello world', multiline=False)
        TextInput:
            id: trrib
            multiline:False
            pos_hint: {'center_x': 0.5, 'center_y': 0.6}
            size_hint: 0.2, 0.05
            hint_text:'rib'
        TextInput:
            id: tramount
            multiline:False
            pos_hint: {'center_x': 0.5, 'center_y': 0.5}
            size_hint: 0.2, 0.05
            hint_text:'amount'
        TextInput:
            id: result
            pos_hint: {'center_x': 0.5, 'center_y': 0.2}
            size_hint: 0.002, 0.005
            hint_text:'amount'
            
        
        
   




<MenuScreen>:
    
    canvas.before:
     
        Rectangle:
            pos: self.pos
            size: self.size
            source: 'banker.png' 
    FloatLayout:
    
        post:0,0
        size: root.width,root.height
        TextInput:
            id: serverid
            multiline:False
            pos_hint: {'center_x': 0.2, 'center_y': 0.9}
            size_hint: 0.2, 0.05
            hint_text:'Server ip'
            
        Button:
            id:connexionbutton
            background_color: 1,0,0,1
            text:'Con'
            size_hint:.04,.02
            pos_hint: {'center_x': 0.05, 'center_y': 0.9}
            on_release:
                root.myconnect(serverid.text)

                

                    
        Button:
            background_color: 0.27, 0.89, 0.43,1
            size_hint:.2,.1
            text:'View Transactions'
            pos_hint:{'x':0.094,'y':0.5}
            text: 'Monitor Account'
            on_release: 
                root.manager.current = 'monitoring'

                
        Button:
            background_color: 0.27, 0.89, 0.43,1
            size_hint:.2,.1
            text:'View Transactions'
            pos_hint:{'x':0.705,'y':0.5}
            text: 'Make Transactions'
            on_release: 
                root.manager.current = 'transactions'

<SettingsScreen>:
    BoxLayout:
        Button:
            text: 'My settings button'
        Button:
            text: 'Back to menu'
            on_press: root.manager.current = 'menu'
""")

SERVER = "192.168.1.11"
PORT = 8080
client = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
# Declare both screens
class MenuScreen(Screen):
    def myconnect(self,ip):
        print(ip)
        
        rsp=""
        try:
            client.connect((ip, PORT))
            client.sendall(bytes("hello",'UTF-8'))
            data = client.recv(3072) 
            rsp=data.decode()         
        except socket.error as e:
            print("socket disconnected")
        if(rsp=="hello"):
            data = client.recv(3072) 
            print("i'm all the way up")
            self.ids.connexionbutton.background_color = (0,1,0,1)
            return True
        else:
            print("failed")
            return False
class MonitoringScreen(Screen):
    def myfunction(self,instance):
        print(f"func_abc: Called from Button with text={instance.text}")


    pass
class TransactionsScreen(Screen):
    def updateResult(self,msg):
        self.ids.result.text= msg
        self.ids.result.size_hint = (0.1,0.1)
    def withdraw(self,trib,tvalue):
        client.sendall(bytes("Withdraw,{},{}".format(trib,tvalue),'UTF-8'))
        data =client.recv(3072)
        rsp=data.decode()
        self.updateResult(rsp)
        pass
    def deposit(self,trib,tvalue):
        client.sendall(bytes("Deposit,{},{}".format(trib,tvalue),'UTF-8'))
        data =client.recv(3072)
        rsp=data.decode()
        self.updateResult(rsp)
        pass
    pass

class InfoScreen(Screen):
    def updatetransaction(self,msg):
        self.ids.transactions.text= msg
        print(msg)
        self.manager.current = 'info'
    def showTransactions(self,ref):
        client.sendall(bytes("ViewTR,{}".format(ref),'UTF-8'))
        data =client.recv(3072)
        rsp=data.decode()
        self.updatetransaction(rsp)
    def showbalance(self,ref):
        client.sendall(bytes("ViewSold,{}".format(ref),'UTF-8'))
        data =client.recv(3072)
        rsp=data.decode()
        self.updatetransaction(rsp)
    def showfacture(self,ref):
        client.sendall(bytes("ViewF,{}".format(ref),'UTF-8'))
        data =client.recv(3072)
        rsp=data.decode()
        self.updatetransaction(rsp)


        pass
class SettingsScreen(Screen):
    

    def func_abc(self, instance):
        print(f"func_abc: Called from Button with text={instance.text}")

    def func_xyz(self, instance):
        print(f"func_xyz: Called from Button with text={instance.text}")
    def ack(self, instance):
        print(f"func_abc: Called from Button with text={instance.text}")
    def showtext(self,text):
        print(text)


        
        client.close()

    def showfacture(self,ref):
          
        print("facture of",ref," is:")

# Create the screen manager
class ScreenManagement(ScreenManager):
    pass


class TheLazyBankerApp(App):

    def build(self):
        return ScreenManagement()
class ribinput(TextInput):
    text="gemme sum tarara:"
if __name__ == '__main__':
    TheLazyBankerApp().run()

clientoob.py

import os
import socket
clear = lambda: os.system('cls')
def viewsold_c(client):
  clear()
  print("enter the rib account to view your sold:",end="")
  rib=input()
  #print("ViewTR,{}".format(rib))
  client.sendall(bytes("ViewSold,{}".format(rib),'UTF-8'))  

def viewtr_c(client):
  clear()
  print("enter the rib account to view your transactions")
  rib=input()
 # print("ViewTR,{}".format(rib))
  client.sendall(bytes("ViewTR,{}".format(rib),'UTF-8'))  

def viewfa_c(client):
  clear()
  print("enter the rib account to view your factures")
  rib=input()
  client.sendall(bytes("ViewF,{}".format(rib),'UTF-8'))  

def transaction_c():
  clear()
  print("enter 1 for making a deposit")
  print("enter 2 for withdrowal")
  print("enter 3 to exit")
  rsp=input()
  while(int(rsp) not in [1,2,3]):
    print("enter a valid option(1, 2 or 3)!")
    rsp=input()
  msg=""
  if int(rsp)==1:
    print("enter your bank account reference")
    rib=input()
    print("enter your amount")
    amount=input()
    msg="Deposit,{},{}".format(rib,amount)
    client.sendall(bytes(msg,'UTF-8'))  

  if int(rsp)==2:
    print("enter your bank account reference")
    rib=input()
    print("enter your amount")
    amount=input()
    msg="Withdraw,{},{}".format(rib,amount)
    client.sendall(bytes(msg,'UTF-8'))  

  if int(rsp)==3:
    queryuser(client)
    
def queryuser(client):
  clear()
  response=0
  print("enter 1 for viewing ur sold")
  print("enter 2 for viewing ur transactions history")
  print("enter 3 for viewing ur factures")
  print("enter 4 for making a transaction")
  print("enter Desired Action:",end="")
  response=input()
  while int(response)not in [1,2,3,4]:
    print("please enter a valid action:")
    response=input()
  if(int(response) ==1):
    viewsold_c(client)
  if(int(response) ==2):
    viewtr_c(client)
  if(int(response) ==3):
    viewfa_c(client)
  if(int(response) ==4):
    transaction_c()
#def withdraw():
#  client.sendall(bytes(out_data,'UTF-8'))  
SERVER = "192.168.56.1"
PORT = 8080
client = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
client.connect((SERVER, PORT))
client.sendall(bytes("hello",'UTF-8'))
in_data =  client.recv(30720)
while True:

  queryuser(client)
  #if out_data=="":
   # while out_data=="":
     # print("Desired action cannot be null!\nenter Desired Action :",end="")
    #  queryuser()
      #out_data = input()
  
  #client.sendall(bytes(out_data,'UTF-8'))
  in_data =  client.recv(5072)
  if(in_data.decode()!="hello"):
    clear()
    print("From Server :" ,in_data.decode())
    input("Press Enter to continue...")
  #if out_data=='bye':
  #  break
  if(in_data.decode()=="bye"):

    break
client.close()


serveroob.py
from fileinput import close
from platform import release
import socket, threading
import math

mahdismutex=threading.Lock()
allowed_actions=[]
allowed_actions.append("ViewSold")
allowed_actions.append("ViewTR")
allowed_actions.append("ViewF")
allowed_actions.append("Deposit")
allowed_actions.append("Withdraw")
current_threads=[]
msgsize=1024
facts="facture.txt"
accs="accounts.txt"
hist="history.txt"
class ClientThread(threading.Thread):
    def __init__(self,clientAddress,clientsocket):
        threading.Thread.__init__(self)
        self.csocket = clientsocket
        print ("New connection added: ", clientAddress)
    def run(self):
        #found the definition for this in here https://docs.python.org/3/library/threading.html#threading.Thread.run    :
        #You may override this method in a subclass. 
        #The standard run() method invokes the callable object passed to the object’s constructor as the target argument, 
        #if any, with positional and keyword arguments taken from the args and kwargs arguments, respectively.
        print ("Connection from : ", clientAddress)
        self.csocket.send(bytes("hello",'utf-8'))
        rsp = ''
        while True:
            try:
                data = self.csocket.recv(3072)          
            except socket.error as e:
                print("socket disconnected")
                break
            rsp = data.decode()
            if rsp!="hello":
                print("communication from client required action:",rsp.split(",")[0])
                keyword=rsp.split(",")[0]
                if keyword in allowed_actions:
                    notify(clientAddress,rsp,self.csocket)
                elif rsp=='bye':
                    break
                #print ("from client:", msg)
                else:
                    msg="Not a known action"
                    self.csocket.send(bytes(msg,'UTF-8'))

        print("Client at ", clientAddress , " disconnected...")   
#end of class
def notify(ip,message,csock):
    elements=message.split(",")
    if elements[0] == "ViewSold":
        msg=GetSold(elements[1])

        csock.send(bytes(msg,'UTF-8'))
    if elements[0] == "ViewTR":
        msg=GetTransactions(elements[1])

        csock.send(bytes(msg,'UTF-8'))
    if elements[0] =="ViewF":
        msg=GetFacture(elements[1])
        csock.send(bytes(msg,'UTF-8'))  
    if elements[0] == "Deposit":
        mahdismutex.acquire()
        if(Deposit(int(elements[1]),int(elements[2]))):
            
            msg="successful deposit"
            csock.send(bytes(msg,'UTF-8'))
        else:
            msg="wrong account reference"
            csock.send(bytes(msg,'UTF-8'))
        mahdismutex.release()
    if elements[0] == "Withdraw":
        mahdismutex.acquire()
        if(Withdraw(int(elements[1]),int(elements[2]))):
            msg="successfull withdrawal"
            csock.send(bytes(msg,'UTF-8'))
        else:
            msg="withdrawal failed"
            csock.send(bytes(msg,'UTF-8'))          
        mahdismutex.release()





def GetSold(ref):
    accounts =open("accounts.txt",'r') 
    ac_list_of_lines = accounts.readlines()
    for i in ac_list_of_lines:
        columns=i.split(',')
        if int(columns[0])==int(ref):
            sign = -1 if columns[2]=="Negativ"  else 1
            accounts.close()
            msg="\nyour solde is:{}".format(int(columns[1])*sign)   
            return  msg
    return "account not found!"

def GetTransactions(ref):
    response=""
    history =open("history.txt",'r') 
    hs_list_of_lines = history.readlines()
    for i in hs_list_of_lines:
        columns=i.split(',')
        if int(columns[0])==int(ref):
            response+="Transaction:\nType:{}   Value:{}   Result:{}   Account-Status-After-Transactions:{}\n".format(columns[1],columns[2],columns[3],columns[4])
    history.close()
    if response=="":
        response="No transactions made under this accounts reference"
    return response

def GetFacture(ref):
    facture =open("facture.txt",'r') 
    fs_list_of_lines = facture.readlines()
    for i in fs_list_of_lines:
        columns=i.split(',')
        if int(columns[0])==int(ref):
            return "Facture to pay next time:"+columns[1]
    return "account not found!"

def AcountExists(ref):
    accounts=open(accs,"r")
    ac_list_of_lines = accounts.readlines()
    facturedamount=0
    for i in range(len(ac_list_of_lines)):
        columns=ac_list_of_lines[i].split(',')
        if int(columns[0])==ref:
            return True
    return False

def UpdateBill(ref,value):#ref mta3 l compte , val hiya li valeur li jbedha
    accounts=open(accs,"r")
    ac_list_of_lines = accounts.readlines()
    facturedamount=0
    for i in range(len(ac_list_of_lines)):
        columns=ac_list_of_lines[i].split(',')
        if int(columns[0])==ref:
            if(columns[2]=="Negativ"):
                facturedamount=value*0.02
            else:
                amount=int(columns[1])
                if ((amount-value)<0 ):
                    facturedamount=-(amount-value)*0.02
                    print("a facture is due")
                    break
                else:
                    break
    facture=open("facture.txt","r")
    fa_list_of_lines = facture.readlines()

    for i in range(len(fa_list_of_lines)):
        columns=fa_list_of_lines[i].split(',')
        if columns[0]=="\n":
            break
        if int(columns[0])==ref:
            print("factured amount:",facturedamount)
            columns[1]=math.trunc(facturedamount + int(columns[1]))
            if (i< len(fa_list_of_lines)-1):
                fa_list_of_lines[i]="{},{}\n".format(columns[0],columns[1])
                facture.close()
            else :
                fa_list_of_lines[i]="{},{}\n".format(columns[0],columns[1])
                facture.close()
            break
    open('facture.txt', 'w').close()
    facture=open("facture.txt","a")
    #PRINT ZAYEDA
    print(fa_list_of_lines)
    for i in fa_list_of_lines:
        facture.write(i)

def Withdraw(ref,amount):
    success=False
    isNegative=False
    if amount<0:
        print("cannot withdraw negativ money nice try XD")
        return False
    if AcountExists(ref):
        accounts=open(accs,"r")
        ac_list_of_lines = accounts.readlines()
        for i in range(len(ac_list_of_lines)):
            columns=ac_list_of_lines[i].split(',')
            if int(columns[0])==ref:
                if(columns[2]=="Negativ"):
                    isNegative=True
                    
                    if (int(columns[1])+amount)<=int(columns[3]):
                        UpdateBill(ref,amount)
                        columns[1]=int(columns[1])+amount
                        ac_list_of_lines[i]="{},{},Negativ,{}".format(columns[0],columns[1],columns[3])
                        accounts.close()
                        success=True
                        break

                        
                if(columns[2]=="Positiv"):
                    if (int(columns[1])-amount)>0:
                        columns[1]=int(columns[1])-amount
                        ac_list_of_lines[i]="{},{},Positiv,{}".format(columns[0],columns[1],columns[3])
                        accounts.close()
                        success=True
                        break
                    elif abs(int(columns[1])-amount)<= int(columns[3]):
                        UpdateBill(ref,amount)
                        ac_list_of_lines[i]="{},{},Negativ,{}".format(columns[0],abs(int(columns[1])-amount),columns[3])
                        accounts.close()
                        success=True
                        isNegative=True
                        break
        open('accounts.txt', 'w').close()  
        accounts=open(accs,"a")
        for i in ac_list_of_lines:
            accounts.write(i)
        history =open(hist,'a') 
        if success:
            if isNegative:
                history.write("\n{},Withdrawal,{},Success,Negativ".format(ref,amount))
            else:
                history.write("\n{},Withdrawal,{},Success,Positiv".format(ref,amount))
        else:
            if isNegative:
                history.write("\n{},Withdrawal,{},Failure,Negativ".format(ref,amount))
            else:
                history.write("\n{},Withdrawal,{},Failure,Positiv".format(ref,amount))
        history.close()
        return success              
    else:
        return False

def Deposit(ref,amount):
    if(AcountExists(ref)):
        #file opened
        facture=open(facts,"r") ##time to pay your debt buddy
        fa_list_of_lines=facture.readlines()
        facture.close()
        accounts=open(accs,"r")
        ac_list_of_lines=accounts.readlines()
        accounts.close()
        history=open(hist,"a")
        for i in range(len(fa_list_of_lines)):
            columns=fa_list_of_lines[i].split(',')
            if int(columns[0])==ref:
                if(int(columns[1])>=amount):
                    columns[1]=int(columns[1])-amount
                    amount=0
                    fa_list_of_lines[i]="{},{}".format(columns[0],columns[1])
                else:
                    amount-=int(columns[1])
                    columns[1]=0
                    if(i!=len(fa_list_of_lines)-1):
                        fa_list_of_lines[i]="{},{}\n".format(columns[0],columns[1])
                    else:
                        fa_list_of_lines[i]="{},{}".format(columns[0],columns[1])

                #updating the accounts after paying debts
                #file opened
                for i in range(len(ac_list_of_lines)):
                    columns=ac_list_of_lines[i].split(',')
                    if int(columns[0])==ref:
                        if(columns[2]=="Negativ"):
                            if  int(columns[1])>amount:
                                columns[1]= int(columns[1])-amount
                                ac_list_of_lines[i]="{},{},{},{}".format(columns[0],columns[1],columns[2],columns[3]) 
                                #156,Withdrawal,100,Success,Negativ format
                                history.write("\n{},Deposit,{},Success,Negativ".format(columns[0],amount)) 
                                
                            else:
                                history.write("\n{},Deposit,{},Success,Positiv".format(columns[0],amount)) 
                                columns[1]= amount-int(columns[1])
                                columns[2]=="Positiv"
                                ac_list_of_lines[i]="{},{},{},{}".format(columns[0],columns[1],"Positiv",columns[3])   
                                

                            
                        else:
                            columns[1]= int(columns[1])+amount
                            history.write("\n{},Deposit,{},Success,Positiv".format(columns[0],amount)) 
                            ac_list_of_lines[i]="{},{},{},{}".format(columns[0],columns[1],columns[2],columns[3])  
                            

        print(ac_list_of_lines)
        history.close()
        open(accs,'w').close()  
        open(facts,'w').close() 
        accounts=open(accs,"a")
        factures=open(facts,"a")
        for i in ac_list_of_lines:
            accounts.write(i)        
        for i in fa_list_of_lines:
            factures.write(i)
        accounts.close()
        factures.close() 
        return True
    else:
        return False       
        
              
LOCALHOST = "192.168.1.8" #whatever the loca ipconfig / ifconfig / ipaddr gives u 
PORT = 8080
server = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
#type=socket.SOCK_DGRAM for udp

server.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)

server.bind((LOCALHOST, PORT))

print("Server is up")
print("Waiting for client request..")
while True:#main boucle XD
    server.listen(1)
    clientsock, clientAddress = server.accept()# returns  a couple (sock,address)
    newthread = ClientThread(clientAddress, clientsock)
    newthread.start()
    current_threads.append(newthread)
